packages/search-in-workspace/src/browser/components/input-with-history.tsx

```diff
@@ -0,0 +1,141 @@
+};
+type InputAttributes = React.InputHTMLAttributes<HTMLInputElement>;
+
+export class InputWithHistory extends React.Component<InputAttributes, HistoryState> {
+    static LIMIT = 100;
+
+    private input = React.createRef<HTMLInputElement>();
+
+    constructor(props: InputAttributes) {
+        super(props);
+        this.state = {
+            history: [],
+            index: 0,
+        };
+    }
+
+    setHistory(history: string[]): void {
+        this.setState(prevState => ({
+            ...prevState,
+            history,
+        }));
+    }
+
+    setIndex(index: number): void {
+        const { history } = this.state;
+        this.value = history[index];
+        this.setState(prevState => ({
+            ...prevState,
+            index,
+        }));
+    }
+
+    get value(): string {
+        return this.input.current?.value ?? '';
+    }
+
+    set value(value: string) {
+        if (this.input.current) {
+            this.input.current.value = value;
+        }
+    }
+
+    /**
+     * Handle history navigation without overriding the parent's onKeyDown handler, if any.
+     */
+    protected readonly onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>): void => {
+        if (Key.ARROW_UP.keyCode === KeyCode.createKeyCode(e.nativeEvent).key?.keyCode) {
+            e.preventDefault();
+            this.previousValue();
+        } else if (Key.ARROW_DOWN.keyCode === KeyCode.createKeyCode(e.nativeEvent).key?.keyCode) {
+            e.preventDefault();
+            this.nextValue();
+        }
+        this.props.onKeyDown?.(e);
+    };
+
+    /**
+     * Switch the input's text to the previous value, if any.
+     */
+    previousValue(): void {
+        const { history, index } = this.state;
+        if (index > 0 && index < history.length) {
+            this.setIndex(index - 1);
+        }
+    }
+
+    /**
+     * Switch the input's text to the next value, if any.
+     */
+    nextValue(): void {
+        const { history, index } = this.state;
+        if (index >= 0 && index < history.length - 1) {
+            this.setIndex(index + 1);
+        }
+    }
+
+    /**
+     * Handle history collection without overriding the parent's onChange handler, if any.
+     */
+    protected readonly onChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+        this.addToHistory();
+        this.props.onChange?.(e);
+    };
+
+    /**
+     * Add a nonempty current value to the history, if not already present. (Debounced, 1 second delay.)
+     */
+    readonly addToHistory = debounce(this.doAddToHistory, 1000);
+
+    private doAddToHistory(): void {
+        if (!this.value) {
+            return;
+        }
+        let { history } = this.state;
+        history = history.filter(term => term !== this.value).concat(this.value);
+        this.reduceToLimit(history);
+        this.setHistory(history);
+        this.setIndex(history.length - 1);
+    }
+
+    private reduceToLimit(history: string[]): string[] {
```

Comment:
You return the history, but in the place where you use this method, you ignore the return, because this method has mutated the input anyway. Is there a use case that calls for the return / did you mean this to mutate?

Reply:
I'll change it to be clearer.